File: src/types/api/provider.ts
/**
 * Provider API type definitions
 * Types for managing model providers in the application
 */

export interface ProviderProxySettings {
  enabled: boolean
  url: string
  username: string
  password: string
  no_proxy: string
  ignore_ssl_certificates: boolean
  proxy_ssl: boolean
  proxy_host_ssl: boolean
  peer_ssl: boolean
  host_ssl: boolean
}

export interface Provider {
  id: string
  name: string
  type: ProviderType
  icon?: string
  enabled: boolean
  api_key?: string
  base_url?: string
  proxy_settings?: ProviderProxySettings
  is_default?: boolean
  created_at?: string
  updated_at?: string
}

export type ProviderType =
  | 'local'
  | 'openai'
  | 'anthropic'
  | 'groq'
  | 'gemini'
  | 'mistral'
  | 'custom'

export interface CreateProviderRequest {
  name: string
  type: ProviderType
  enabled?: boolean
  api_key?: string
  base_url?: string
}

export interface UpdateProviderRequest {
  name?: string
  enabled?: boolean
  api_key?: string
  base_url?: string
  proxy_settings?: ProviderProxySettings
}

export interface ProviderListResponse {
  providers: Provider[]
  total: number
  page: number
  per_page: number
}

export interface TestProviderProxyResponse {
  success: boolean
  message: string
}

// Device detection types
export interface DeviceInfo {
  id: number
  name: string
  device_type: string // cpu, cuda, metal
  memory_total?: number // Total memory in bytes
  memory_free?: number // Free memory in bytes
  is_available: boolean
}

export interface AvailableDevicesResponse {
  devices: DeviceInfo[]
  default_device_type: string
  supports_multi_gpu: boolean
}

File: src-tauri/src/database/models/provider.rs
use chrono::{DateTime, Utc};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use super::{model::DeviceType, proxy::ProxySettings};
use crate::database::macros::impl_string_to_enum;
use crate::database::types::JsonOption;

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, sqlx::Type)]
#[serde(rename_all = "lowercase")]
pub enum ProviderType {
    Local,
    OpenAI,
    Anthropic,
    Groq,
    Gemini,
    Mistral,
    DeepSeek,
    Huggingface,
    Custom,
}

impl ProviderType {
    pub fn from_str(s: &str) -> Option<Self> {
        match s {
            "local" => Some(ProviderType::Local),
            "openai" => Some(ProviderType::OpenAI),
            "anthropic" => Some(ProviderType::Anthropic),
            "groq" => Some(ProviderType::Groq),
            "gemini" => Some(ProviderType::Gemini),
            "mistral" => Some(ProviderType::Mistral),
            "deepseek" => Some(ProviderType::DeepSeek),
            "huggingface" => Some(ProviderType::Huggingface),
            "custom" => Some(ProviderType::Custom),
            _ => Some(ProviderType::Custom), // fallback to custom for unknown types
        }
    }

    pub fn as_str(&self) -> &'static str {
        match self {
            ProviderType::Local => "local",
            ProviderType::OpenAI => "openai",
            ProviderType::Anthropic => "anthropic",
            ProviderType::Groq => "groq",
            ProviderType::Gemini => "gemini",
            ProviderType::Mistral => "mistral",
            ProviderType::DeepSeek => "deepseek",
            ProviderType::Huggingface => "huggingface",
            ProviderType::Custom => "custom",
        }
    }
}

// Implement string to enum conversion for SQLx
impl_string_to_enum!(ProviderType);

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct Provider {
    pub id: Uuid,
    pub name: String,
    #[serde(rename = "type")]
    pub provider_type: ProviderType,
    pub enabled: bool,
    pub api_key: Option<String>,
    pub base_url: Option<String>,
    pub built_in: bool,
    pub proxy_settings: JsonOption<ProxySettings>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct CreateProviderRequest {
    pub name: String,
    pub enabled: Option<bool>,
    pub api_key: Option<String>,
    pub base_url: Option<String>,
    pub proxy_settings: Option<ProxySettings>,
    #[serde(rename = "type")]
    pub provider_type: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct UpdateProviderRequest {
    pub name: Option<String>,
    pub enabled: Option<bool>,
    pub api_key: Option<String>,
    pub base_url: Option<String>,
    pub proxy_settings: Option<ProxySettings>,
}

// Device detection structures
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct DeviceInfo {
    pub id: i32, // Device index (0, 1, 2, etc.)
    pub name: String,
    pub device_type: DeviceType,
    pub memory_total: Option<u64>, // Total memory in bytes
    pub memory_free: Option<u64>,  // Free memory in bytes
    pub is_available: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct AvailableDevicesResponse {
    pub devices: Vec<DeviceInfo>,
    pub default_device_type: DeviceType,
    pub supports_multi_gpu: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct ProviderListResponse {
    pub providers: Vec<Provider>,
    pub total: i64,
    pub page: i32,
    pub per_page: i32,
}
